### ISSUE TITLE
[2020 NeurIPS] Energy-based Out-of-distribution Detection

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#62]
기존의 energy score을 이용한 방식들은 likelihood score log p(x)를 generative model을 통해 사용하였는데 이는 실제로 optimize하는데 intractable + unstable

해당 논문은 discriminative classification model만을 이용하여 energy score을 계산하며, 이를 이용하여 OOD detection을 할 수 있으며 뛰어남을 보임

## 1. Background : Energy-based Models

Energy-based model (EBM)은 input space를 single, non-probabilistic scalar인 energy로 mapping해주는 함수임

<img width="893" alt="image" src="https://user-images.githubusercontent.com/49643709/163115003-7596a98f-aa8e-47c9-958c-eb8003dc900b.png">

> - Gibbs distribution = Boltzmann distribution
> - [wiki link](https://en.wikipedia.org/wiki/Boltzmann_distribution)

다양한 input이 들어오면 그에 해당되는 energy values의 집합이 존재하게 될텐데, Gibbs distribution을 통해 이를 probability density로 바꿔줄 수 있음

<img width="901" alt="image" src="https://user-images.githubusercontent.com/49643709/163115293-83c9d6ca-e03f-4bd7-ab34-82b2b7445dd8.png">

> - F : Helmholtz free energy
> - k : Boltzmann constant
> - T : absolute temperature
> - Z : partition function
> - [wiki link](https://en.wikipedia.org/wiki/Helmholtz_free_energy)

어떠한 시스템이 constant volume, temperature, and particle number을 유지하는 경우, 이를 canonical ensemble이라고 말하며 canonical ensemble일 때 Helmholtz free energy를 partition function으로 나타낼 수 있음

이를 단순화하여 다음과 같이 표현할 수 있음

<img width="711" alt="image" src="https://user-images.githubusercontent.com/49643709/163117632-72951bd7-3722-4cfe-8148-0ad471d91133.png">

해당 식은 discriminative neural classifier의 softmax function을 이용한 categorical distribution과 연결지을 수 있음

<img width="711" alt="image" src="https://user-images.githubusercontent.com/49643709/163117700-8e4557b4-f635-4f3f-a43e-b146363ced36.png">

Equation 1, 3을 연결하여 neural network의 free energy function을 다음과 같이 표현할 수 있음

<img width="715" alt="image" src="https://user-images.githubusercontent.com/49643709/163117763-b6c8a185-f785-46a4-9fa5-9109d8460edf.png">

즉 logit 값들의 -log sum exp를 통해 모델 f에 대한 input x의 free energy (scalar)를 구할 수 있음

참고로 E(x, y) = - f_y(x)로 class y에 대한 energy를 의미하며 E(x ; f)는 모델 f에 대한 input x의 free energy를 의미함

## 2. Energy-based Out-of-distribution Detection

논문에서 energy-based OOD detection 방법 2가지를 제시함

1. Using energy as an OOD score for pre-trained models
2. Use energy as a trainable cost function for model fine-tuning

### 2.1. Energy as Inference-time OOD Score
<img width="722" alt="image" src="https://user-images.githubusercontent.com/49643709/163212916-acfbcf96-ad43-4454-be48-3e21e076e44f.png">

Model이 negative log-likelihood (NLL) loss로 학습하면 in-distribution data point의 energy를 push down함을 보임
(NLL loss가 0이 되려면 E(x, y)는 0이 되어야하며 E(x, j)는 값이 커져야함)

<img width="730" alt="image" src="https://user-images.githubusercontent.com/49643709/163202288-fb6a9af6-6d66-4a5f-9144-ce245a99ece7.png">

In-distribution의 energy는 낮고 OOD의 energy는 높음

Positive samples는 high scores를 가진다는 conventional definition과 energy score를 align하기 위해 negative energy scores를 사용

즉 logit에 log sum exp operator를 적용하면 됨

### 2.2. Energy Score vs Softmax Score
<img width="713" alt="image" src="https://user-images.githubusercontent.com/49643709/163204409-2ffa233b-ad97-4cbb-9ce1-8a1330650d98.png">

<img width="331" alt="image" src="https://user-images.githubusercontent.com/49643709/163203026-9e0bfd22-b252-41f0-ae61-973dc58f517e.png">

Max softmax probability (MSP)는 결국 p(y|x)의 max 값을 이용하여 OOD detection을 함

양변에 log를 취해 softmax score를 free energy function에 대한 식으로 표현하여 전개할 수 있음
(즉 MSP는 energy score의 특별한 case임)

전개식에서 충돌하는 부분이 있는데 모델이 nll loss로 학습하게 되면 free energy는 in-distribution dataset에 대해 감소하도록 학습하는데 max logit value는 값이 커지도록 학습됨

이렇게 반대되는 방향성으로 인해 MSP를 biased scoring function이라고 할 수 있으며 OOD detection에 적합하지 않음을 보임

### 2.3. Energy-bounded Learning for OOD Detection
<img width="722" alt="image" src="https://user-images.githubusercontent.com/49643709/163204522-be4877e3-5c28-447c-afa1-f22c922c3d4f.png">

Energy-bounded learning objective를 통해 neural network를 fine-tune하여 in-distribution data는 lower energy를, OOD data는 higher energy를 갖도록 학습을 강제할 수 있음

논문은 two squared hinge loss terms를 통해 in-distribution의 energy는 m_in보다 낮게, OOD data는 m_out보다 크게 만드는 regularization loss를 추가하여 학습함
(1개의 hyperparameter를 가지는 hinge loss를 사용해 학습하게 되면 hyperparameter의 value가 stabilize하지 않았다고 함)

### 3.1. Setup
<img width="724" alt="image" src="https://user-images.githubusercontent.com/49643709/163208249-da730ae1-763f-494b-b149-fbe432e1259d.png">

- OOD dataset의 pixel values는 z-normalization을 통해 normalize
- Auxiliary outlier dataset으로 80 Million Tiny Images 사용
- L_energy weight hyperparameter lambda는 0.1로 설정하여 사용

### 3.2. Energy-based vs Softmax-based
<img width="727" alt="image" src="https://user-images.githubusercontent.com/49643709/163205867-ee1f855f-99ec-4ab8-ae1c-fbeb88d3fbe7.png">

<img width="733" alt="image" src="https://user-images.githubusercontent.com/49643709/163206089-50d6443d-f40e-424f-a3b4-9bca79388c76.png">

- Energy-based가 더 좋음

### 3.3. How does our approach compare to competitive OOD detection methods?
<img width="737" alt="image" src="https://user-images.githubusercontent.com/49643709/163207092-db859044-4af1-46c0-8d29-42a17900b030.png">

<img width="663" alt="image" src="https://user-images.githubusercontent.com/49643709/163207279-600aaa09-f421-45cf-8eb5-2758ad9ec7b1.png">

- 대부분의 방법들은 pre-trained model을 어느정도 tune해야하거나 classifier가 구분할 수 있게 additional data를 필요로 함
- Pre-trained network에서의 energy score 방식은 parameter-free, easy to use and deploy한다는 장점이 있으며, 성능 면에서 충분히 경쟁력 있
- Generative model 기반의 OOD detection과 비교했을 때, energy-based method가 다른 방법들보다 성능이 좋음

### 3.4. Ablation : Temperature Scaling & Margin Parameter
<img width="719" alt="image" src="https://user-images.githubusercontent.com/49643709/163207529-58758195-f432-4c54-9cfa-ce2b9d0ebd9e.png">

- Using larger T leads to more uniformly distributed predictions and makes the energy scores less distinguishable between in- and out-of-distribution examples
- m_out은 not very sensitive, m_in 값을 너무 작게 설정하면 optimization이 어려워지고 performance degradation 일어남

### 3.5. Does energy fine-tuning affect the classification accuracy?
<img width="729" alt="image" src="https://user-images.githubusercontent.com/49643709/163207925-e6a1e381-bba8-45cb-b3bc-84c97753da4e.png">

- Fine-tuning을 하면 확실히 OOD detection performance는 올라가며, classification accuracy는 충분히 유의미하게 유지함 (almost comparable)
